aardvark/retrievers/runner.py

```diff
@@ -0,0 +1,261 @@
+import asyncio
+import logging
+import re
+from copy import copy
+from typing import Any, Dict, List
+
+import confuse
+from asgiref.sync import sync_to_async
+from botocore.exceptions import ClientError
+from cloudaux.aws.iam import list_roles, list_users
+from cloudaux.aws.sts import boto3_cached_conn
+from swag_client import InvalidSWAGDataException
+from swag_client.backend import SWAGManager
+from swag_client.util import parse_swag_config_options
+
+from aardvark.exceptions import RetrieverException
+from aardvark.plugins import AardvarkPlugin
+from aardvark.persistence.sqlalchemy import SQLAlchemyPersistence
+from aardvark.retrievers import RetrieverPlugin
+from aardvark.retrievers.access_advisor import AccessAdvisorRetriever
+
+log = logging.getLogger("aardvark")
+sap = SQLAlchemyPersistence()
+re_account_id = re.compile(r"\d{12}")
+
+
+class RetrieverRunner(AardvarkPlugin):
+    """Scheduling and execution for data retrieval tasks."""
+
+    retrievers: List[RetrieverPlugin]
+    account_queue: asyncio.Queue
+    arn_queue: asyncio.Queue
+    results_queue: asyncio.Queue
+    failed_arns: List[str]
+    tasks: List[asyncio.Future]
+    num_workers: int
+    swag: SWAGManager
+    swag_config: confuse.ConfigView
+
+    def __init__(
+        self,
+        alternative_config: confuse.Configuration = None,
+    ):
+        super().__init__(alternative_config=alternative_config)
+        self.tasks = []
+        self.retrievers = []
+        self.failed_arns = []
+        self.num_workers = self.config["updater"]["num_threads"].as_number()
+        self.swag_config = self.config["swag"]
+        swag_opts = parse_swag_config_options(self.swag_config["opts"].get())
+        self.swag = SWAGManager(**swag_opts)
+
+    def register_retriever(self, r: RetrieverPlugin):
+        """Add a retriever instance to be called during the run process."""
+        self.retrievers.append(r)
+
+    async def _run_retrievers(self, arn: str) -> Dict[str, Any]:
+        """Run retriever plugins for a given ARN.
+
+        Retriever plugins are executed in the order in which they are registered. Each retriever
+        is passed the result from the previous one, starting with a dict containing an `arn` element.
+
+        Note: The data from all previous retriever plugins is mutable by subsequent ones.
+        """
+        data = {
+            "arn": arn,
+        }
+        # Iterate through retrievers, passing the results from the previous to the next.
```

Comment:
So retrievers are acting as compounding gathering and enrichment pipelines?

Reply:
Will this play nicely with the Aardvark Digest pattern we are looking into?

Reply:
Yeah, that's a good way to describe them. Subsequent retrievers can do whatever they want to the data dict that's passed in to them (for better or worse). So my thought is for the action usage retriever to enrich the Access Advisor data in a similar way to how action-level data is returned from AA for S3 (i.e. in the nested `TrackedActionsLastAccessed` list for each service).